### DISCUSSION TITLE
Implement a get by path with default value

### DISCUSSION THREAD

**Discussion by alvarolb** [nlohmann/json#2585] [Ideas]
I use this library quite extensively inside a large project, and I found that I normally require to access to nested paths of an object, where many times the value is missing, or the content is not what expected. Doing that with the current library can be complicated and iterative, so I ended up doing a small function to look for a key in a path and return it, or return a default value if not found, not of the same type, etc.

The code is the following:

```cpp
template <class T>
static T get_value(const nlohmann::json& json, const std::string& path, T value = T(), size_t max_deep=25){
    // discard if empty or not an object,
    if(json.is_null() || json.empty() || !json.is_object()) return value;

    std::stringstream test(path);
    std::string key_name;
    auto current = std::ref(json);
    auto index = 0;

    while(std::getline(test, key_name, '.') && index<max_deep)
    {
        nlohmann::json::const_iterator it = current.get().find(key_name);
        if(it!=current.get().end()){
            current = it.value();
            ++index;
        }else{
            return value;
        }
    }

    try{
        T returnValue = current.get();
        return returnValue;
    }catch(const std::exception& e){
        return value;
    }catch(...){
        return value;
    }
}
```

I use like that: 

```cpp
std::string value = get_value<std::string>(json, "user.email"); // will get ["user"]["email"] in string or empty string
int value = get_value<std::string>(json, "user.attempts", 0); 
```
And it saves a lot of code checking for json root type, nested objects, value type, etc. It is a first attempt and for sure it can be improved a lot, but just wanted to share if someone find it useful, or something similar can be included in the library.

Thanks! Regards.

**Comment by gregmarr** [comment]
Have you looked at the JSON Pointer support?  I haven't used it myself, but it looks like this does what you want:

    auto value = json.value("/user/attempts"_json_pointer, std::string{});
    auto int = json.value("/user/attempts"_json_pointer, 0);

https://github.com/nlohmann/json/blob/master/doc/examples/basic_json__value_ptr.cpp
